### ISSUE TITLE
Augment target dependency / application order

### ISSUE THREAD

**Issue by ddjoh** [orchestron-orchestrator/acton-yang#2]
To handle augments upon augments we need to apply them in dependency order.

For this we could either build a dependency graph
  or
we could consider whether simply ordering the application of augments by their path-length (by number of nodes) is more efficient while still *complete*?

**Comment by plajjan** [comment]
I believe you encountered this in the real world, right? What model was that? We need a test case, preferably a minimal one, which I suppose we'll have to write ourselves, but I want to be sure we don't miss any nuance in the model you've seen.

As for the dependency graph, I believe we need both. We currently don't consider module dependencies before compiling them. We definitely need to but we also need to order augments within a module since they aren't specified in any particular order.

I'll try to take a quick stab at this and see how it goes, will report back shortly :)

**Comment by ddjoh** [comment]
Cheers!

`ietf-network-instance.yang` applied an augment `/if:interfaces/if:interface/ip:ipv4` that was augmented by `ietf-ip.yang` onto `ietf-interfaces`.

**Comment by plajjan** [comment]
@ddjoh I started with the intra-module dependencies and implemented your suggestion of ordering by target path length. Works for my simple example. I tried to think if there are complex uses of like groupings / uses that then has augments where we might end up with the wrong order... and I don't know. Can't immediately think of an example that breaks this.

Thanks for the ietf example. That's different modules though, so needs a different implementation!
